lib/core/src/Cardano/Wallet/Primitive/CoinSelection/Collateral.hs

```diff
@@ -0,0 +1,404 @@
+        -- as collateral.
+    , minimumSelectionAmount
+        :: Coin
+        -- ^ A lower bound on the sum of coins to be selected as collateral.
+    , searchSpaceLimit
+        :: SearchSpaceLimit
+        -- ^ An upper bound on the search space size, to protect the wallet
+        -- against computations that use excessive amounts of time or space.
+    }
+    deriving (Eq, Generic, Show)
+
+-- | Specifies an upper bound on the search space size.
+--
+data SearchSpaceLimit
+    = SearchSpaceLimit Int
+    -- ^ Specifies an upper bound on the number of coin combinations that can
+    -- be considered in any single step.
+    | UnsafeNoSearchSpaceLimit
+    -- ^ Specifies that there is no search space limit. This should only be
+    -- used for testing purposes.
+    deriving (Eq, Show)
+
+-- | Represents a successful selection of collateral.
+--
+newtype SelectCollateralResult inputId = SelectCollateralResult
+    { coinsSelected :: Map inputId Coin
+        -- ^ The coins that were selected for collateral.
+    }
+    deriving (Eq, Generic, Show)
+
+-- | Represents an unsuccessful attempt to select collateral.
+--
+newtype SelectCollateralError inputId = SelectCollateralError
+    { largestCombinationAvailable :: Map inputId Coin
+        -- ^ The largest combination of coins available.
+    }
+    deriving (Eq, Generic, Show)
+
+-- | Selects coins for collateral.
+--
+-- This function tries two strategies in the following order, picking the first
+-- strategy that succeeds:
+--
+--    1. Attempt to select an amount of collateral that is as small as possible.
+--    2. Attempt to select collateral from the largest coins available.
+--
+-- The first strategy, given unlimited computation time, will always produce an
+-- optimal result: the smallest possible amount of collateral. However, if the
+-- required search space is large, and if the 'searchSpaceLimit' parameter is
+-- set to a value that's smaller than the required search space size, then this
+-- strategy will fail without computing a result.
+--
+-- The second strategy sacrifices optimality and always produces a result if
+-- one is available, by looking only at the very largest coins available. This
+-- result can be computed very quickly, without using much search space.
+--
+-- The combination of these two strategies means that we can satisfy the
+-- following properties:
+--
+-- If the attempt to select collateral succeeds:
+--
+--    >>> sum  coinsSelected ≥ minimumSelectionAmount
+--    >>> size coinsSelected ≤ maximumSelectionSize
+--    >>>      coinsSelected ⊆ coinsAvailable
+--
+-- If the attempt to select collateral fails:
+--
+--    >>> sum  largestCombinationAvailable < minimumSelectionAmount
+--    >>> size largestCombinationAvailable ≤ maximumSelectionSize
+--    >>>      largestCombinationAvailable ⊆ coinsAvailable
+--
+selectCollateral :: forall inputId. Ord inputId => SelectCollateral inputId
```

Comment:
This is marked by `weeder` as a weed. We could solve this by adding a basic test for it.

Reply:
Fixed in d88cddfdeb1f4dd88b1cb64a2b7c59303aab0dce.